File: src/main/scala/org/joshjoyce/stingray/Normal.scala
package org.joshjoyce.stingray

import org.joshjoyce.stingray._

object Normal {
  def apply(p: Point3d): Normal = Normal(p.x, p.y, p.z)
  def apply(v: Vector3d): Normal = Normal(v.x, v.y, v.z)
}

case class Normal(x: Real, y: Real, z: Real) {
  def +(n: Normal) = Normal(x + n.x, y + n.y, z + n.z)
  
  def +(v: Vector3d) = Vector3d(x + v.x, y + v.y, z + v.z)

  def dot(n: Normal) = n.x * x + n.y * y + n.z * z
  
  def *(a: Real) = Normal(a * x, a * y, a * z)

File: src/main/java/org/joshjoyce/sunbeam/Normal.java
package org.joshjoyce.sunbeam;

public class Normal {
    public final double x;
    public final double y;
    public final double z;

    public static Normal fromPoint3d(Point3d p) {
        return new Normal(p.x, p.y, p.z);
    }

    public static Normal fromVector3d(Vector3d v) {
        return new Normal(v.x, v.y, v.z);
    }

    public Normal(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Normal add(Normal n) {
        return new Normal(x + n.x, y + n.y, z + n.z);
    }

    public Normal add(Vector3d v) {
        return new Normal(x + v.x, y + v.y, z + v.z);
    }

    public double dot(Normal n) {
        return x * n.x + y * n.y + z * n.z;
    }

    public Normal scale(double d) {
        return new Normal(d * x, d * y, d * z);
    }
}
